pyomo/contrib/mindtpy/algorithm_base_class.py

```diff
@@ -0,0 +1,2280 @@
+                                'of {}'.format(subprob_terminate_cond))
+            self.should_terminate = True
+            self.results.solver.status = SolverStatus.error
+    
+    ######################################################################################################################################################
+    # iterate.py
+
+
+    def algorithm_should_terminate(self, config, check_cycling):
+        """Checks if the algorithm should terminate at the given point.
+
+        This function determines whether the algorithm should terminate based on the solver options and progress.
+        (Sets the self.results.solver.termination_condition to the appropriate condition, i.e. optimal,
+        maxIterations, maxTimeLimit).
+
+        Parameters
+        ----------
+        config : ConfigBlock
+            The specific configurations for MindtPy.
+        check_cycling : bool
+            Whether to check for a special case that causes the discrete variables to loop through the same values.
+
+        Returns
+        -------
+        bool
+            True if the algorithm should terminate, False otherwise.
+        """
+        if self.should_terminate:
+            if self.primal_bound == self.primal_bound_progress[0]:
+                self.results.solver.termination_condition = tc.noSolution
+            else:
+                self.results.solver.termination_condition = tc.feasible
+            return True
+
+        # Check bound convergence
+        if self.abs_gap <= config.absolute_bound_tolerance:
+            config.logger.info(
+                'MindtPy exiting on bound convergence. '
+                'Absolute gap: {} <= absolute tolerance: {} \n'.format(
+                    self.abs_gap, config.absolute_bound_tolerance))
+            self.results.solver.termination_condition = tc.optimal
+            return True
+        # Check relative bound convergence
+        if self.best_solution_found is not None:
+            if self.rel_gap <= config.relative_bound_tolerance:
+                config.logger.info(
+                    'MindtPy exiting on bound convergence. '
+                    'Relative gap : {} <= relative tolerance: {} \n'.format(
+                        self.rel_gap, config.relative_bound_tolerance))
+
+        # Check iteration limit
+        if self.mip_iter >= config.iteration_limit:
+            config.logger.info(
+                'MindtPy unable to converge bounds '
+                'after {} main iterations.'.format(self.mip_iter))
+            config.logger.info(
+                'Final bound values: Primal Bound: {}  Dual Bound: {}'.
+                format(self.primal_bound, self.dual_bound))
+            if config.single_tree:
+                self.results.solver.termination_condition = tc.feasible
+            else:
+                self.results.solver.termination_condition = tc.maxIterations
```

Comment:
This is because `mip_iter` equals `1` in single-tree implementation. So single-tree implementation will not terminate by iteration limit. It's a little bit wired.

Reply:
Is there any way of unifying these? Maybe not for this PR, but later we can go back to he drawing board and think how can we better integrate the single-tree and multi-tree methods